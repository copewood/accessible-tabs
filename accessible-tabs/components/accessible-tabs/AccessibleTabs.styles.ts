import styled from 'styled-components';

export const StyledTabs = styled.div`
    fontFamily: 'lucida grande', sans-serif;
    color: grey;
`;

export const StyledTabList = styled.div`
    min-width: 550px;
`;

export const StyledTab = styled.button`
    &[role="tab"], [role="tab"]:focus, [role="tab"]:hover {
        color: blue;
        position: relative;
        z-index: 2;
        top: 2px;
        margin: 0;
        margin-top: 4px;
        padding: 3px 3px 4px;
        border: 1px solid hsl(219deg 1% 72%);
        border-bottom: 2px solid hsl(219deg 1% 72%);
        border-radius: 5px 5px 0 0;
        overflow: visible;
        background: hsl(220deg 20% 94%);
        outline: none;
        font-weight: bold;
    }
    &[role="tab"][aria-selected="true"] {
        padding: 2px 2px 4px;
        margin-top: 0;
        border-width: 2px;
        border-top-width: 6px;
        border-top-color: rgb(36 116 214);
        border-bottom-color: hsl(220deg 43% 99%);
        background: hsl(220deg 43% 99%);
    }
    &[role="tab"][aria-selected="false"] {
        border-bottom: 1px solid hsl(219deg 1% 72%);
    }
    &[role="tab"] span.focus {
        display: inline-block;
        margin: 2px;
        padding: 4px 6px;
    }
    &[role="tab"]:hover span.focus,
    &[role="tab"]:focus span.focus,
    &[role="tab"]:active span.focus {
        padding: 2px 4px;
        border: 2px solid rgb(36 116 214);
        border-radius: 3px;
    }
`;

export const StyledTabPanel = styled.div`
    &[role="tabpanel"] {
        padding: 5px;
        border: 2px solid hsl(219deg 1% 72%);
        border-radius: 0 5px 5px;
        background: hsl(220deg 43% 99%);
        min-height: 10em;
        min-width: 550px;
        overflow: auto;
    }
    &[role="tabpanel"].is-hidden {
        display: none;
    }
    &[role="tabpanel"] p {
        margin: 0;
    }
    & a {
        text-decoration: underline;
        color: blue;
        margin-right: .5rem;
    }
`;  