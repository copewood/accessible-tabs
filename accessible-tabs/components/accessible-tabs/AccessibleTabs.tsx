import React, { FC, useEffect } from 'react';

import { TabsManual } from '../tabs-manual/TabsManual';

import sanitizeHtml from 'sanitize-html';

import { 
  StyledTabs,
  StyledTabList,
  StyledTab,
  StyledTabPanel
} from './AccessibleTabs.styles';

type Tab = {
  name: string;
  link: string;
  fullName: string;
  bio: string;
}

type AccessibleTabsProps = {
  title: string;
  tabsIndex: number;
  tabs: Array<Tab>;
};

export const AccessibleTabs: FC<AccessibleTabsProps> = ({title, tabsIndex, tabs}) => {

  // Initialize tablist
  useEffect(() => {
      var tablists = document.querySelectorAll('[role=tablist].manual');
      for (var i = 0; i < tablists.length; i++) {
        new TabsManual(tablists[i]);
      }
  }, [])

  return <>
    <StyledTabs>
      <h3 id={`tablist-${tabsIndex}`}>
        {title}
      </h3>

      <StyledTabList 
          key={`tablist-${tabsIndex}`} 
          role="tablist"
          aria-labelledby={`tablist-${tabsIndex}`}
          className="manual">

        {tabs.map((tab, index) => 
          <StyledTab 
            id={`tablist-${tabsIndex}-tab-${index}`}
            key={`tablist-${tabsIndex}-tab-${index}`}
            type="button"
            role="tab"
            aria-selected={index === 0 ? 'true' : 'false'}
            aria-controls={`tablist-${tabsIndex}-tabpanel-${index}`}>
            <span className="focus">
              {tab.name}
            </span>
          </StyledTab>
        )}
      </StyledTabList>

      {tabs.map((tab, index) => {
        const cleanBio = sanitizeHtml(tab.bio);
        return (
          <StyledTabPanel 
            id={`tablist-${tabsIndex}-tabpanel-${index}`}
            key={`tablist-${tabsIndex}-tabpanel-${index}`}
            role="tabpanel"
            aria-labelledby={`tablist-${tabsIndex}-tab-${index}`}>
            <p>
              <a href={tab.link}>
                {tab.fullName}
              </a>
              <span dangerouslySetInnerHTML={
                {__html: cleanBio}
              }/>
            </p>
          </StyledTabPanel>
        )
      })}
    </StyledTabs>
  </>;
};
