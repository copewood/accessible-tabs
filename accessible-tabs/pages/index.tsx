import Head from 'next/head';
import styled from 'styled-components';

import testData from 'data/test-data.json';

import { AccessibleTabs } from '@components/accessible-tabs/AccessibleTabs';

export const Container = styled.div`
  min-height: 100vh;
  padding: 0 0.5rem;
  display: flex;
  flex-direction: column;
  justify-content: start;
  align-items: center;
  height: 100vh;
`;

export const Main = styled.div`
  padding: 1rem 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: start;
  align-items: center;
`;

export const Title = styled.div`
  margin: 0;
  line-height: 1.15;
  font-size: 1.5rem;
  text-align: center;
  a {
    color: #0070f3;
    text-decoration: underline;
    margin: 1rem;
    font-size: 1rem;
  }
`;

export default function Home() {
  return (
    <Container>
      <Head>
        <title>
          Next.js, TypeScript, Styled-Components, Jest, Testing-Library
        </title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <Main>
        <Title>
          Iress Accessible Tabs Coding Test
          <div>
            {testData.map((testItem, tabsIndex) =>  
               testItem.tabs.map((tab, index) => 
                <a 
                  href={`#tablist-${tabsIndex}-tab-${index}`}
                  key={`#tablist-${tabsIndex}-link-${index}`}
                >
                  {tab.name}
                </a>
              ))}
          </div>
        </Title>
        
        {testData.map((testItem, tabsIndex) =>
          <AccessibleTabs
            title={testItem.title}
            tabsIndex={tabsIndex}
            tabs={testItem.tabs}
          />
        )}

      </Main>
    </Container>
  );
}
