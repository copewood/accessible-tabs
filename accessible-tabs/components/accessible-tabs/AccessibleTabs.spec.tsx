import React from 'react';
import { render, screen } from '@testing-library/react';

import testData from '../../data/test-data.json';

import { AccessibleTabs } from '../../components/accessible-tabs/AccessibleTabs';

describe('Accessible Tabs', () => {

  it('should renderd Accessible Tabs', () => {
    const { container } = render(
      <AccessibleTabs 
        title={testData[0].title}
        tabsIndex={0}
        tabs={testData[0].tabs} 
      />
    );
    expect(container).toMatchSnapshot()
  });
});
