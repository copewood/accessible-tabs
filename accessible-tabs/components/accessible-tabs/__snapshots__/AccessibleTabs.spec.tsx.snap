// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accessible Tabs should renderd Accessible Tabs 1`] = `
<div>
  <div
    class="sc-bczRLJ Kvmjt"
  >
    <h3
      id="tablist-0"
    >
      Danish Composers
    </h3>
    <div
      aria-labelledby="tablist-0"
      class="sc-gsnTZi oDuzQ manual"
      role="tablist"
    >
      <button
        aria-controls="tablist-0-tabpanel-0"
        aria-selected="true"
        class="sc-dkzDqf kSATmS"
        id="tablist-0-tab-0"
        role="tab"
        type="button"
      >
        <span
          class="focus"
        >
          Maria Ahlefeldt
        </span>
      </button>
      <button
        aria-controls="tablist-0-tabpanel-1"
        aria-selected="false"
        class="sc-dkzDqf kSATmS"
        id="tablist-0-tab-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="focus"
        >
          Carl Andersen
        </span>
      </button>
      <button
        aria-controls="tablist-0-tabpanel-2"
        aria-selected="false"
        class="sc-dkzDqf kSATmS"
        id="tablist-0-tab-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="focus"
        >
          Ida da Fonseca
        </span>
      </button>
      <button
        aria-controls="tablist-0-tabpanel-3"
        aria-selected="false"
        class="sc-dkzDqf kSATmS"
        id="tablist-0-tab-3"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="focus"
        >
          Peter Lange-Müller
        </span>
      </button>
    </div>
    <div
      aria-labelledby="tablist-0-tab-0"
      class="sc-hKMtZM jMZDUb"
      id="tablist-0-tabpanel-0"
      role="tabpanel"
    >
      <p>
        <a
          href="https://en.wikipedia.org/wiki/Maria_Theresia_Ahlefeldt"
        >
          Maria Theresia Ahlefeldt
        </a>
        <span>
          (16 January 1755 – 20 December 1810) was a Danish, (originally German), composer. She is known as the first female composer in Denmark. Maria Theresia composed music for several ballets, operas, and plays of the royal theatre. She was given good critic as a composer and described as a "
          <span>
            virkelig Tonekunstnerinde
          </span>
          " ('a True Artist of Music').
        </span>
      </p>
    </div>
    <div
      aria-labelledby="tablist-0-tab-1"
      class="sc-hKMtZM jMZDUb is-hidden"
      id="tablist-0-tabpanel-1"
      role="tabpanel"
    >
      <p>
        <a
          href="https://en.wikipedia.org/wiki/Joachim_Andersen_(composer)"
        >
          Carl Joachim Andersen
        </a>
        <span>
          (29 April 1847 – 7 May 1909) was a Danish flutist, conductor and composer born in Copenhagen, son of the flutist Christian Joachim Andersen. Both as a virtuoso and as composer of flute music, he is considered one of the best of his time. He was considered to be a tough leader and teacher and demanded as such a lot from his orchestras but through that style he reached a high level.
        </span>
      </p>
    </div>
    <div
      aria-labelledby="tablist-0-tab-2"
      class="sc-hKMtZM jMZDUb is-hidden"
      id="tablist-0-tabpanel-2"
      role="tabpanel"
    >
      <p>
        <a
          href="https://en.wikipedia.org/wiki/Ida_Henriette_da_Fonseca"
        >
          Ida Henriette da Fonseca
        </a>
        <span>
          (July 27, 1802 – July 6, 1858) was a Danish opera singer and composer.  Ida Henriette da Fonseca was the daughter of Abraham da Fonseca (1776–1849) and Marie Sofie Kiærskou (1784–1863). She and her sister Emilie da Fonseca were students of Giuseppe Siboni, choir master of the Opera in Copenhagen. She was given a place at the royal Opera alongside her sister the same year she debuted in 1827.
        </span>
      </p>
    </div>
    <div
      aria-labelledby="tablist-0-tab-3"
      class="sc-hKMtZM jMZDUb is-hidden"
      id="tablist-0-tabpanel-3"
      role="tabpanel"
    >
      <p>
        <a
          href="https://en.wikipedia.org/wiki/Peter_Lange-M%C3%BCller"
        >
          Peter Erasmus Lange-Müller
        </a>
        <span>
          (1 December 1850 – 26 February 1926) was a Danish composer and pianist. His compositional style was influenced by Danish folk music and by the work of Robert Schumann; Johannes Brahms; and his Danish countrymen, including J.P.E. Hartmann.
        </span>
      </p>
    </div>
  </div>
</div>
`;
